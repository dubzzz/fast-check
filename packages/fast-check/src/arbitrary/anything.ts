import type { Arbitrary } from '../check/arbitrary/definition/Arbitrary';
import { anyArbitraryBuilder } from './_internals/builders/AnyArbitraryBuilder';
import type { ObjectConstraints } from './_internals/helpers/QualifiedObjectConstraints';
import { toQualifiedObjectConstraints } from './_internals/helpers/QualifiedObjectConstraints';

export type { ObjectConstraints };

/**
 * For any type of values
 *
 * You may use {@link sample} to preview the values that will be generated
 *
 * @example
 * ```javascript
 * null, undefined, 42, 6.5, 'Hello', {}, {k: [{}, 1, 2]}
 * ```
 *
 * @remarks Since 0.0.7
 * @public
 */
function anything(): Arbitrary<unknown>;
/**
 * For any type of values following the constraints defined by `settings`
 *
 * You may use {@link sample} to preview the values that will be generated
 *
 * @example
 * ```javascript
 * null, undefined, 42, 6.5, 'Hello', {}, {k: [{}, 1, 2]}
 * ```
 *
 * @example
 * ```typescript
 * // Using custom settings
 * fc.anything({
 *     key: fc.char(),
 *     values: [fc.integer(10,20), fc.constant(42)],
 *     maxDepth: 2
 * });
 * // Can build entries such as:
 * // - 19
 * // - [{"2":12,"k":15,"A":42}]
 * // - {"4":[19,13,14,14,42,11,20,11],"6":42,"7":16,"L":10,"'":[20,11],"e":[42,20,42,14,13,17]}
 * // - [42,42,42]...
 * ```
 *
 * @param constraints - Constraints to apply when building instances
 *
 * @remarks Since 0.0.7
 * @public
 */
function anything(constraints: ObjectConstraints): Arbitrary<unknown>;
function anything(constraints?: ObjectConstraints): Arbitrary<unknown> {
  return anyArbitraryBuilder(toQualifiedObjectConstraints(constraints));
}
export { anything };
