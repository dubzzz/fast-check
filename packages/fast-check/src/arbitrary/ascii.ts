import type { Arbitrary } from '../check/arbitrary/definition/Arbitrary';
import { buildCharacterArbitrary } from './_internals/builders/CharacterArbitraryBuilder';
import { indexToPrintableIndexMapper, indexToPrintableIndexUnmapper } from './_internals/mappers/IndexToPrintableIndex';

/**
 * For single ascii characters - char code between 0x00 (included) and 0x7f (included)
 * @remarks Since 0.0.1
 * @public
 */
export function ascii(): Arbitrary<string> {
  return buildCharacterArbitrary(0x00, 0x7f, indexToPrintableIndexMapper, indexToPrintableIndexUnmapper);
}
