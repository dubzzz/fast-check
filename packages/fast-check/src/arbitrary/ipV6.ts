import { array } from './array';
import type { Arbitrary } from '../check/arbitrary/definition/Arbitrary';
import { oneof } from './oneof';
import { hexaString } from './hexaString';
import { tuple } from './tuple';
import { ipV4 } from './ipV4';
import {
  fullySpecifiedMapper,
  fullySpecifiedUnmapper,
  onlyTrailingMapper,
  onlyTrailingUnmapper,
  multiTrailingMapper,
  multiTrailingUnmapper,
  multiTrailingMapperOne,
  multiTrailingUnmapperOne,
  singleTrailingMapper,
  singleTrailingUnmapper,
  noTrailingMapper,
  noTrailingUnmapper,
} from './_internals/mappers/EntitiesToIPv6';

/** @internal */
function h16sTol32Mapper([a, b]: [string, string]): string {
  return `${a}:${b}`;
}

/** @internal */
function h16sTol32Unmapper(value: unknown): [string, string] {
  if (typeof value !== 'string') throw new Error('Invalid type');
  if (!value.includes(':')) throw new Error('Invalid value');
  return value.split(':', 2) as [string, string];
}

/**
 * For valid IP v6
 *
 * Following {@link https://tools.ietf.org/html/rfc3986#section-3.2.2 | RFC 3986}
 *
 * @remarks Since 1.14.0
 * @public
 */
export function ipV6(): Arbitrary<string> {
  // h16 = 1*4HEXDIG
  // ls32 = ( h16 ":" h16 ) / IPv4address
  // IPv6address   =                            6( h16 ":" ) ls32
  //               /                       "::" 5( h16 ":" ) ls32
  //               / [               h16 ] "::" 4( h16 ":" ) ls32
  //               / [ *1( h16 ":" ) h16 ] "::" 3( h16 ":" ) ls32
  //               / [ *2( h16 ":" ) h16 ] "::" 2( h16 ":" ) ls32
  //               / [ *3( h16 ":" ) h16 ] "::"    h16 ":"   ls32
  //               / [ *4( h16 ":" ) h16 ] "::"              ls32
  //               / [ *5( h16 ":" ) h16 ] "::"              h16
  //               / [ *6( h16 ":" ) h16 ] "::"

  // Any size-based arbitrary called within the implementation of ipV6 has
  // to be called with size:'max' in order to prevent any behaviour change
  // related to global settings on size. ipV6 is fully independent of size
  const h16Arb = hexaString({ minLength: 1, maxLength: 4, size: 'max' });
  const ls32Arb = oneof(tuple(h16Arb, h16Arb).map(h16sTol32Mapper, h16sTol32Unmapper), ipV4());
  return oneof(
    tuple(array(h16Arb, { minLength: 6, maxLength: 6, size: 'max' }), ls32Arb).map(
      fullySpecifiedMapper,
      fullySpecifiedUnmapper,
    ),
    tuple(array(h16Arb, { minLength: 5, maxLength: 5, size: 'max' }), ls32Arb).map(
      onlyTrailingMapper,
      onlyTrailingUnmapper,
    ),
    tuple(
      array(h16Arb, { minLength: 0, maxLength: 1, size: 'max' }),
      array(h16Arb, { minLength: 4, maxLength: 4, size: 'max' }),
      ls32Arb,
    ).map(multiTrailingMapper, multiTrailingUnmapper),
    tuple(
      array(h16Arb, { minLength: 0, maxLength: 2, size: 'max' }),
      array(h16Arb, { minLength: 3, maxLength: 3, size: 'max' }),
      ls32Arb,
    ).map(multiTrailingMapper, multiTrailingUnmapper),
    tuple(
      array(h16Arb, { minLength: 0, maxLength: 3, size: 'max' }),
      array(h16Arb, { minLength: 2, maxLength: 2, size: 'max' }),
      ls32Arb,
    ).map(multiTrailingMapper, multiTrailingUnmapper),
    tuple(array(h16Arb, { minLength: 0, maxLength: 4, size: 'max' }), h16Arb, ls32Arb).map(
      multiTrailingMapperOne,
      multiTrailingUnmapperOne,
    ),
    tuple(array(h16Arb, { minLength: 0, maxLength: 5, size: 'max' }), ls32Arb).map(
      singleTrailingMapper,
      singleTrailingUnmapper,
    ),
    tuple(array(h16Arb, { minLength: 0, maxLength: 6, size: 'max' }), h16Arb).map(
      singleTrailingMapper,
      singleTrailingUnmapper,
    ),
    tuple(array(h16Arb, { minLength: 0, maxLength: 7, size: 'max' })).map(noTrailingMapper, noTrailingUnmapper),
  );
}
