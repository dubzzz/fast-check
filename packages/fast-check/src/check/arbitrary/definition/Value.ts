import type { WithCloneMethod } from '../../symbols';
import { cloneMethod, hasCloneMethod } from '../../symbols';

const safeObjectDefineProperty = Object.defineProperty;

/**
 * A `Value<T, TShrink = T>` holds an internal value of type `T`
 * and its associated context
 *
 * @remarks Since 3.0.0 (previously called `NextValue` in 2.15.0)
 * @public
 */
export class Value<T> {
  /**
   * State storing the result of hasCloneMethod
   * If `true` the value will be cloned each time it gets accessed
   * @remarks Since 2.15.0
   */
  readonly hasToBeCloned: boolean;
  /**
   * Flag indicating whether or not the this.value has already been called once
   * If so, the underlying will be cloned
   * Only set when hasToBeCloned = true
   */
  private readOnce: boolean;
  /**
   * Safe value of the shrinkable
   * Depending on `hasToBeCloned` it will either be `value_` or a clone of it
   * @remarks Since 2.15.0
   */
  readonly value!: T;
  /**
   * Internal value of the shrinkable
   * @remarks Since 2.15.0
   */
  readonly value_: T;
  /**
   * Context for the generated value
   * TODO - Do we want to clone it too?
   * @remarks 2.15.0
   */
  readonly context: unknown;

  /**
   * @param value_ - Internal value of the shrinkable
   * @param context - Context associated to the generated value (useful for shrink)
   * @param customGetValue - Limited to internal usages (to ease migration to next), it will be removed on next major
   */
  constructor(value_: T, context: unknown, customGetValue: (() => T) | undefined = undefined) {
    this.value_ = value_;
    this.context = context;
    this.hasToBeCloned = customGetValue !== undefined || hasCloneMethod(value_);
    this.readOnce = false;
    if (this.hasToBeCloned) {
      safeObjectDefineProperty(this, 'value', { get: customGetValue !== undefined ? customGetValue : this.getValue });
    } else {
      this.value = value_;
    }
  }

  /** @internal */
  private getValue() {
    if (this.hasToBeCloned) {
      if (!this.readOnce) {
        this.readOnce = true;
        return this.value_;
      }
      return (this.value_ as unknown as WithCloneMethod<T>)[cloneMethod]();
    }
    return this.value_;
  }
}
