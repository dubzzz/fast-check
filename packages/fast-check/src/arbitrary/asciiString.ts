import { Arbitrary } from '../check/arbitrary/definition/Arbitrary';
import { array } from './array';
import { ascii } from './ascii';
import { StringSharedConstraints } from './_shared/StringSharedConstraints';
import { codePointsToStringMapper, codePointsToStringUnmapper } from './_internals/mappers/CodePointsToString';
export { StringSharedConstraints } from './_shared/StringSharedConstraints';

/**
 * For strings of {@link ascii}
 *
 * @param constraints - Constraints to apply when building instances (since 2.4.0)
 *
 * @remarks Since 0.0.1
 * @public
 */
export function asciiString(constraints: StringSharedConstraints = {}): Arbitrary<string> {
  return array(ascii(), constraints).map(codePointsToStringMapper, codePointsToStringUnmapper);
}
