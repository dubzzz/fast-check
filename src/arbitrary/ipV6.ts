import { array } from './array';
import { Arbitrary } from '../check/arbitrary/definition/Arbitrary';
import { oneof } from './oneof';
import { hexaString } from './hexaString';
import { tuple } from './tuple';
import { ipV4 } from './ipV4';
import {
  fullySpecifiedMapper,
  fullySpecifiedUnmapper,
  onlyTrailingMapper,
  onlyTrailingUnmapper,
  multiTrailingMapper,
  multiTrailingUnmapper,
  multiTrailingMapperOne,
  multiTrailingUnmapperOne,
  singleTrailingMapper,
  singleTrailingUnmapper,
  noTrailingMapper,
  noTrailingUnmapper,
} from './_internals/mappers/EntitiesToIPv6';

/** @internal */
function h16sTol32Mapper([a, b]: [string, string]): string {
  return `${a}:${b}`;
}

/** @internal */
function h16sTol32Unmapper(value: unknown): [string, string] {
  if (typeof value !== 'string') throw new Error('Invalid type');
  if (!value.includes(':')) throw new Error('Invalid value');
  return value.split(':', 2) as [string, string];
}

/**
 * For valid IP v6
 *
 * Following {@link https://tools.ietf.org/html/rfc3986#section-3.2.2 | RFC 3986}
 *
 * @remarks Since 1.14.0
 * @public
 */
export function ipV6(): Arbitrary<string> {
  // h16 = 1*4HEXDIG
  // ls32 = ( h16 ":" h16 ) / IPv4address
  // IPv6address   =                            6( h16 ":" ) ls32
  //               /                       "::" 5( h16 ":" ) ls32
  //               / [               h16 ] "::" 4( h16 ":" ) ls32
  //               / [ *1( h16 ":" ) h16 ] "::" 3( h16 ":" ) ls32
  //               / [ *2( h16 ":" ) h16 ] "::" 2( h16 ":" ) ls32
  //               / [ *3( h16 ":" ) h16 ] "::"    h16 ":"   ls32
  //               / [ *4( h16 ":" ) h16 ] "::"              ls32
  //               / [ *5( h16 ":" ) h16 ] "::"              h16
  //               / [ *6( h16 ":" ) h16 ] "::"
  const h16Arb = hexaString({ minLength: 1, maxLength: 4 });
  const ls32Arb = oneof(tuple(h16Arb, h16Arb).map(h16sTol32Mapper, h16sTol32Unmapper), ipV4());
  return oneof(
    tuple(array(h16Arb, { minLength: 6, maxLength: 6 }), ls32Arb).map(fullySpecifiedMapper, fullySpecifiedUnmapper),
    tuple(array(h16Arb, { minLength: 5, maxLength: 5 }), ls32Arb).map(onlyTrailingMapper, onlyTrailingUnmapper),
    tuple(array(h16Arb, { minLength: 0, maxLength: 1 }), array(h16Arb, { minLength: 4, maxLength: 4 }), ls32Arb).map(
      multiTrailingMapper,
      multiTrailingUnmapper
    ),
    tuple(array(h16Arb, { minLength: 0, maxLength: 2 }), array(h16Arb, { minLength: 3, maxLength: 3 }), ls32Arb).map(
      multiTrailingMapper,
      multiTrailingUnmapper
    ),
    tuple(array(h16Arb, { minLength: 0, maxLength: 3 }), array(h16Arb, { minLength: 2, maxLength: 2 }), ls32Arb).map(
      multiTrailingMapper,
      multiTrailingUnmapper
    ),
    tuple(array(h16Arb, { minLength: 0, maxLength: 4 }), h16Arb, ls32Arb).map(
      multiTrailingMapperOne,
      multiTrailingUnmapperOne
    ),
    tuple(array(h16Arb, { minLength: 0, maxLength: 5 }), ls32Arb).map(singleTrailingMapper, singleTrailingUnmapper),
    tuple(array(h16Arb, { minLength: 0, maxLength: 6 }), h16Arb).map(singleTrailingMapper, singleTrailingUnmapper),
    tuple(array(h16Arb, { minLength: 0, maxLength: 7 })).map(noTrailingMapper, noTrailingUnmapper)
  );
}
