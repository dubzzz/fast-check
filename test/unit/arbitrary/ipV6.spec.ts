import { ipV6 } from '../../../src/arbitrary/ipV6';

import { Value } from '../../../src/check/arbitrary/definition/Value';
import {
  assertProduceValuesShrinkableWithoutContext,
  assertShrinkProducesSameValueWithoutInitialContext,
  assertProduceCorrectValues,
  assertProduceSameValueGivenSameSeed,
} from './__test-helpers__/ArbitraryAssertions';
import { buildShrinkTree, renderTree } from './__test-helpers__/ShrinkTree';

describe('ipV6 (integration)', () => {
  const isValidIpV4 = (value: string) => {
    const chunks = value.split('.').map((v) => {
      if (v[0] === '0') {
        if (v[1] === 'x' || v[1] === 'X') return parseInt(v, 16);
        return parseInt(v, 8);
      }
      return parseInt(v, 10);
    });

    // one invalid chunk
    if (chunks.find((v) => Number.isNaN(v)) !== undefined) return false;

    // maximal amount of 4 chunks
    if (chunks.length > 4) return false;

    // all chunks, except the last one are inferior or equal to 255
    if (chunks.slice(0, -1).find((v) => v < 0 && v > 255) !== undefined) return false;

    // last chunk must be below 256^(5 âˆ’ number of chunks)
    return chunks[chunks.length - 1] < 256 ** (5 - chunks.length);
  };
  const isCorrect = (value: string) => {
    const firstElision = value.indexOf('::');
    if (firstElision !== -1) {
      // At most one '::'
      if (value.substr(firstElision + 1).includes('::')) return false;
    }
    const chunks = value.split(':');
    const last = chunks[chunks.length - 1];
    // The ipv4 can only be composed of 4 decimal chunks separated by dots
    // 1.1000 is not a valid IP v4 in the context of IP v6
    const endByIpV4 = last.includes('.') && isValidIpV4(last);

    const nonEmptyChunks = chunks.filter((c) => c !== '');
    const hexaChunks = endByIpV4 ? nonEmptyChunks.slice(0, nonEmptyChunks.length - 1) : nonEmptyChunks;
    if (!hexaChunks.every((s) => /^[0-9a-f]{1,4}$/.test(s))) return false;

    const equivalentChunkLength = endByIpV4 ? hexaChunks.length + 2 : hexaChunks.length;
    return firstElision !== -1 ? equivalentChunkLength < 8 : equivalentChunkLength === 8;
  };

  const ipV6Builder = () => ipV6();

  it('should produce the same values given the same seed', () => {
    assertProduceSameValueGivenSameSeed(ipV6Builder);
  });

  it('should only produce correct values', () => {
    assertProduceCorrectValues(ipV6Builder, isCorrect);
  });

  it('should produce values seen as shrinkable without any context', () => {
    assertProduceValuesShrinkableWithoutContext(ipV6Builder);
  });

  it('should be able to shrink to the same values without initial context', () => {
    assertShrinkProducesSameValueWithoutInitialContext(ipV6Builder);
  });

  it.each`
    source
    ${'::1'}
    ${'0123:5678:9abc:ef01:2345:6789:128.0.0.1'}
    ${'0123:5678:9abc:ef01:2345:6789:0000:ffff'}
    ${'0:56:9abc:ef01:234:67:8.8.8.8'}
    ${'0:56:9abc:ef01:234:67:0:f'}
    ${'::5678:9abc:ef01:2345:6789:128.0.0.1'}
    ${'::5678:9abc:ef01:2345:6789:0000:ffff'}
    ${'::9abc:ef01:2345:6789:128.0.0.1'}
    ${'::9abc:ef01:2345:6789:0000:ffff'}
    ${'5678::9abc:ef01:2345:6789:128.0.0.1'}
    ${'5678::9abc:ef01:2345:6789:0000:ffff'}
    ${'::ef01:2345:6789:128.0.0.1'}
    ${'::ef01:2345:6789:0000:ffff'}
    ${'9abc::ef01:2345:6789:128.0.0.1'}
    ${'9abc::ef01:2345:6789:0000:ffff'}
    ${'5678:9abc::ef01:2345:6789:128.0.0.1'}
    ${'5678:9abc::ef01:2345:6789:0000:ffff'}
    ${'::2345:6789:128.0.0.1'}
    ${'::2345:6789:0000:ffff'}
    ${'ef01::2345:6789:128.0.0.1'}
    ${'ef01::2345:6789:0000:ffff'}
    ${'9abc:ef01::2345:6789:128.0.0.1'}
    ${'9abc:ef01::2345:6789:0000:ffff'}
    ${'5678:9abc:ef01::2345:6789:128.0.0.1'}
    ${'5678:9abc:ef01::2345:6789:0000:ffff'}
    ${'::6789:128.0.0.1'}
    ${'::6789:0000:ffff'}
    ${'2345::6789:128.0.0.1'}
    ${'2345::6789:0000:ffff'}
    ${'ef01:2345::6789:128.0.0.1'}
    ${'ef01:2345::6789:0000:ffff'}
    ${'9abc:ef01:2345::6789:128.0.0.1'}
    ${'9abc:ef01:2345::6789:0000:ffff'}
    ${'5678:9abc:ef01:2345::6789:128.0.0.1'}
    ${'5678:9abc:ef01:2345::6789:0000:ffff'}
    ${'::128.0.0.1'}
    ${'::0000:ffff'}
    ${'2345::128.0.0.1'}
    ${'2345::0000:ffff'}
    ${'ef01:2345::128.0.0.1'}
    ${'ef01:2345::0000:ffff'}
    ${'9abc:ef01:2345::128.0.0.1'}
    ${'9abc:ef01:2345::0000:ffff'}
    ${'5678:9abc:ef01:2345::128.0.0.1'}
    ${'5678:9abc:ef01:2345::0000:ffff'}
    ${'0123:5678:9abc:ef01:2345::128.0.0.1'}
    ${'0123:5678:9abc:ef01:2345::0000:ffff'}
    ${'::0123'}
    ${'6789::0123'}
    ${'2345:6789::0123'}
    ${'ef01:2345:6789::0123'}
    ${'9abc:ef01:2345:6789::0123'}
    ${'5678:9abc:ef01:2345:6789::0123'}
    ${'0123:5678:9abc:ef01:2345:6789::0123'}
    ${'::'}
    ${'0123::'}
    ${'6789:0123::'}
    ${'2345:6789:0123::'}
    ${'ef01:2345:6789:0123::'}
    ${'9abc:ef01:2345:6789:0123::'}
    ${'5678:9abc:ef01:2345:6789:0123::'}
    ${'0123:5678:9abc:ef01:2345:6789:0123::'}
  `('should be able to generate $source with fc.ipV6()', ({ source }) => {
    // Arrange / Act
    const arb = ipV6();
    const out = arb.canShrinkWithoutContext(source);

    // Assert
    expect(out).toBe(true);
  });

  it.each`
    rawValue
    ${'::1'}
    ${'0123:5678:9abc:ef01:2345:6789:128.0.0.1'}
  `('should be able to shrink $rawValue', ({ rawValue }) => {
    // Arrange
    const arb = ipV6();
    const value = new Value(rawValue, undefined);

    // Act
    const renderedTree = renderTree(buildShrinkTree(arb, value, { numItems: 100 })).join('\n');

    // Assert
    expect(arb.canShrinkWithoutContext(rawValue)).toBe(true);
    expect(renderedTree).toMatchSnapshot();
  });
});
