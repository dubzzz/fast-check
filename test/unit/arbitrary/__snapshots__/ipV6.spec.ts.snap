// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ipV6 (integration) should be able to shrink ::1 1`] = `
"\\"::1\\"
└> \\"::0\\""
`;

exports[`ipV6 (integration) should be able to shrink 0123:5678:9abc:ef01:2345:6789:128.0.0.1 1`] = `
"\\"0123:5678:9abc:ef01:2345:6789:128.0.0.1\\"
├> \\"3:5678:9abc:ef01:2345:6789:128.0.0.1\\"
|  ├> \\"0:5678:9abc:ef01:2345:6789:128.0.0.1\\"
|  |  ├> \\"0:5678:9abc:ef01:2345:6789:0.0.0.1\\"
|  |  |  └> \\"0:5678:9abc:ef01:2345:6789:0.0.0.0\\"
|  |  ├> \\"0:5678:9abc:ef01:2345:6789:64.0.0.1\\"
|  |  |  ├> \\"0:5678:9abc:ef01:2345:6789:32.0.0.1\\"
|  |  |  |  ├> \\"0:5678:9abc:ef01:2345:6789:16.0.0.1\\"
|  |  |  |  |  ├> \\"0:5678:9abc:ef01:2345:6789:8.0.0.1\\"
|  |  |  |  |  |  ├> \\"0:5678:9abc:ef01:2345:6789:4.0.0.1\\"
|  |  |  |  |  |  |  ├> \\"0:5678:9abc:ef01:2345:6789:2.0.0.1\\"
|  |  |  |  |  |  |  |  ├> \\"0:5678:9abc:ef01:2345:6789:1.0.0.1\\"
|  |  |  |  |  |  |  |  |  ├> \\"0:5678:9abc:ef01:2345:6789:0.0.0.1\\"
|  |  |  |  |  |  |  |  |  |  └> \\"0:5678:9abc:ef01:2345:6789:0.0.0.0\\"
|  |  |  |  |  |  |  |  |  └> \\"0:5678:9abc:ef01:2345:6789:1.0.0.0\\"
|  |  |  |  |  |  |  |  |     └> \\"0:5678:9abc:ef01:2345:6789:0.0.0.0\\"
|  |  |  |  |  |  |  |  └> \\"0:5678:9abc:ef01:2345:6789:2.0.0.0\\"
|  |  |  |  |  |  |  |     └> \\"0:5678:9abc:ef01:2345:6789:1.0.0.0\\"
|  |  |  |  |  |  |  |        └> \\"0:5678:9abc:ef01:2345:6789:0.0.0.0\\"
|  |  |  |  |  |  |  ├> \\"0:5678:9abc:ef01:2345:6789:3.0.0.1\\"
|  |  |  |  |  |  |  |  ├> \\"0:5678:9abc:ef01:2345:6789:2.0.0.1\\"
|  |  |  |  |  |  |  |  |  ├> \\"0:5678:9abc:ef01:2345:6789:0.0.0.1\\"
|  |  |  |  |  |  |  |  |  |  └> \\"0:5678:9abc:ef01:2345:6789:0.0.0.0\\"
|  |  |  |  |  |  |  |  |  ├> \\"0:5678:9abc:ef01:2345:6789:1.0.0.1\\"
|  |  |  |  |  |  |  |  |  |  ├> \\"0:5678:9abc:ef01:2345:6789:0.0.0.1\\"
|  |  |  |  |  |  |  |  |  |  |  └> \\"0:5678:9abc:ef01:2345:6789:0.0.0.0\\"
|  |  |  |  |  |  |  |  |  |  └> \\"0:5678:9abc:ef01:2345:6789:1.0.0.0\\"
|  |  |  |  |  |  |  |  |  |     └> \\"0:5678:9abc:ef01:2345:6789:0.0.0.0\\"
|  |  |  |  |  |  |  |  |  └> \\"0:5678:9abc:ef01:2345:6789:2.0.0.0\\"
|  |  |  |  |  |  |  |  |     ├> \\"0:5678:9abc:ef01:2345:6789:0.0.0.0\\"
|  |  |  |  |  |  |  |  |     └> \\"0:5678:9abc:ef01:2345:6789:1.0.0.0\\"
|  |  |  |  |  |  |  |  |        └> \\"0:5678:9abc:ef01:2345:6789:0.0.0.0\\"
|  |  |  |  |  |  |  |  └> \\"0:5678:9abc:ef01:2345:6789:3.0.0.0\\"
|  |  |  |  |  |  |  |     └> \\"0:5678:9abc:ef01:2345:6789:2.0.0.0\\"
|  |  |  |  |  |  |  |        ├> \\"0:5678:9abc:ef01:2345:6789:0.0.0.0\\"
|  |  |  |  |  |  |  |        └> \\"0:5678:9abc:ef01:2345:6789:1.0.0.0\\"
|  |  |  |  |  |  |  |           └> \\"0:5678:9abc:ef01:2345:6789:0.0.0.0\\"
|  |  |  |  |  |  |  └> \\"0:5678:9abc:ef01:2345:6789:4.0.0.0\\"
|  |  |  |  |  |  |     ├> \\"0:5678:9abc:ef01:2345:6789:2.0.0.0\\"
|  |  |  |  |  |  |     |  └> \\"0:5678:9abc:ef01:2345:6789:1.0.0.0\\"
|  |  |  |  |  |  |     |     └> \\"0:5678:9abc:ef01:2345:6789:0.0.0.0\\"
|  |  |  |  |  |  |     └> \\"0:5678:9abc:ef01:2345:6789:3.0.0.0\\"
|  |  |  |  |  |  |        └> \\"0:5678:9abc:ef01:2345:6789:2.0.0.0\\"
|  |  |  |  |  |  |           ├> \\"0:5678:9abc:ef01:2345:6789:0.0.0.0\\"
|  |  |  |  |  |  |           └> \\"0:5678:9abc:ef01:2345:6789:1.0.0.0\\"
|  |  |  |  |  |  |              └> \\"0:5678:9abc:ef01:2345:6789:0.0.0.0\\"
|  |  |  |  |  |  ├> \\"0:5678:9abc:ef01:2345:6789:6.0.0.1\\"
|  |  |  |  |  |  |  ├> \\"0:5678:9abc:ef01:2345:6789:5.0.0.1\\"
|  |  |  |  |  |  |  |  ├> \\"0:5678:9abc:ef01:2345:6789:4.0.0.1\\"
|  |  |  |  |  |  |  |  |  ├> \\"0:5678:9abc:ef01:2345:6789:0.0.0.1\\"
|  |  |  |  |  |  |  |  |  |  └> \\"0:5678:9abc:ef01:2345:6789:0.0.0.0\\"
|  |  |  |  |  |  |  |  |  ├> \\"0:5678:9abc:ef01:2345:6789:2.0.0.1\\"
|  |  |  |  |  |  |  |  |  |  ├> \\"0:5678:9abc:ef01:2345:6789:1.0.0.1\\"
|  |  |  |  |  |  |  |  |  |  |  ├> \\"0:5678:9abc:ef01:2345:6789:0.0.0.1\\"
|  |  |  |  |  |  |  |  |  |  |  |  └> \\"0:5678:9abc:ef01:2345:6789:0.0.0.0\\"
|  |  |  |  |  |  |  |  |  |  |  └> \\"0:5678:9abc:ef01:2345:6789:1.0.0.0\\"
|  |  |  |  |  |  |  |  |  |  |     └> \\"0:5678:9abc:ef01:2345:6789:0.0.0.0\\"
|  |  |  |  |  |  |  |  |  |  └> \\"0:5678:9abc:ef01:2345:6789:2.0.0.0\\"
|  |  |  |  |  |  |  |  |  |     └> \\"0:5678:9abc:ef01:2345:6789:1.0.0.0\\"
|  |  |  |  |  |  |  |  |  |        └> \\"0:5678:9abc:ef01:2345:6789:0.0.0.0\\"
|  |  |  |  |  |  |  |  |  ├> \\"0:5678:9abc:ef01:2345:6789:3.0.0.1\\"
|  |  |  |  |  |  |  |  |  |  ├> \\"0:5678:9abc:ef01:2345:6789:2.0.0.1\\"
|  |  |  |  |  |  |  |  |  |  |  ├> \\"0:5678:9abc:ef01:2345:6789:0.0.0.1\\"
|  |  |  |  |  |  |  |  |  |  |  |  └> \\"0:5678:9abc:ef01:2345:6789:0.0.0.0\\"
|  |  |  |  |  |  |  |  |  |  |  ├> \\"0:5678:9abc:ef01:2345:6789:1.0.0.1\\"
|  |  |  |  |  |  |  |  |  |  |  |  ├> \\"0:5678:9abc:ef01:2345:6789:0.0.0.1\\"
|  |  |  |  |  |  |  |  |  |  |  |  |  └> \\"0:5678:9abc:ef01:2345:6789:0.0.0.0\\"
|  |  |  |  |  |  |  |  |  |  |  |  └> \\"0:5678:9abc:ef01:2345:6789:1.0.0.0\\"
|  |  |  |  |  |  |  |  |  |  |  |     └> \\"0:5678:9abc:ef01:2345:6789:0.0.0.0\\"
|  |  |  |  |  |  |  |  |  |  |  └> \\"0:5678:9abc:ef01:2345:6789:2.0.0.0\\"
|  |  |  |  |  |  |  |  |  |  |     ├> \\"0:5678:9abc:ef01:2345:6789:0.0.0.0\\"
|  |  |  |  |  |  |  |  |  |  |     └> \\"0:5678:9abc:ef01:2345:6789:1.0.0.0\\"
|  |  |  |  |  |  |  |  |  |  |        └> \\"0:5678:9abc:ef01:2345:6789:0.0.0.0\\"
|  |  |  |  |  |  |  |  |  |  └> \\"0:5678:9abc:ef01:2345:6789:3.0.0.0\\"
|  |  |  |  |  |  |  |  |  |     └> \\"0:5678:9abc:ef01:2345:6789:2.0.0.0\\"
|  |  |  |  |  |  |  |  |  |        ├> \\"0:5678:9abc:ef01:2345:6789:0.0.0.0\\"
|  |  |  |  |  |  |  |  |  |        └> \\"0:5678:9abc:ef01:2345:6789:1.0.0.0\\"
|  |  |  |  |  |  |  |  |  |           └> \\"0:5678:9abc:ef01:2345:6789:0.0.0.0\\"
|  |  |  |  |  |  |  |  |  └> \\"0:5678:9abc:ef01:2345:6789:4.0.0.0\\"
|  |  |  |  |  |  |  |  |     ├> \\"0:5678:9abc:ef01:2345:6789:0.0.0.0\\"
|  |  |  |  |  |  |  |  |     ├> \\"0:5678:9abc:ef01:2345:6789:2.0.0.0\\"
|  |  |  |  |  |  |  |  |     |  └> \\"0:5678:9abc:ef01:2345:6789:1.0.0.0\\"
|  |  |  |  |  |  |  |  |     |     └> \\"0:5678:9abc:ef01:2345:6789:0.0.0.0\\"
|  |  |  |  |  |  |  |  |     └> \\"0:5678:9abc:ef01:2345:6789:3.0.0.0\\"
|  |  |  |  |  |  |  |  |        └> \\"0:5678:9abc:ef01:2345:6789:2.0.0.0\\"
|  |  |  |  |  |  |  |  |           ├> \\"0:5678:9abc:ef01:2345:6789:0.0.0.0\\"
|  |  |  |  |  |  |  |  |           └> \\"0:5678:9abc:ef01:2345:6789:1.0.0.0\\"
|  |  |  |  |  |  |  |  |              └> \\"0:5678:9abc:ef01:2345:6789:0.0.0.0\\"
|  |  |  |  |  |  |  |  └> \\"0:5678:9abc:ef01:2345:6789:5.0.0.0\\"
|  |  |  |  |  |  |  |     └> \\"0:5678:9abc:ef01:2345:6789:4.0.0.0\\"
|  |  |  |  |  |  |  |        ├> \\"0:5678:9abc:ef01:2345:6789:0.0.0.0\\"
|  |  |  |  |  |  |  |        ├> \\"0:5678:9abc:ef01:2345:6789:2.0.0.0\\"
|  |  |  |  |  |  |  |        |  └> \\"0:5678:9abc:ef01:2345:6789:1.0.0.0\\"
|  |  |  |  |  |  |  |        |     └> \\"0:5678:9abc:ef01:2345:6789:0.0.0.0\\"
|  |  |  |  |  |  |  |        └> \\"0:5678:9abc:ef01:2345:6789:3.0.0.0\\"
|  |  |  |  |  |  |  |           └> \\"0:5678:9abc:ef01:2345:6789:2.0.0.0\\"
|  |  |  |  |  |  |  |              ├> \\"0:5678:9abc:ef01:2345:6789:0.0.0.0\\"
|  |  |  |  |  |  |  |              └> \\"0:5678:9abc:ef01:2345:6789:1.0.0.0\\"
|  |  |  |  |  |  |  |                 └> \\"0:5678:9abc:ef01:2345:6789:0.0.0.0\\"
|  |  |  |  |  |  |  └> \\"0:5678:9abc:ef01:2345:6789:6.0.0.0\\"
|  |  |  |  |  |  |     └> …
|  |  |  |  |  |  └> …
|  |  |  |  |  └> …
|  |  |  |  └> …
|  |  |  └> …
|  |  └> …
|  └> …
└> …"
`;
